File: library/gsheets.py
"""
library/gsheets.py — Shared Google Sheets sync for all job application scripts.

Setup (one-time):
  1. Go to console.cloud.google.com → New project → Enable "Google Sheets API"
  2. IAM & Admin → Service Accounts → Create → Download JSON key
  3. Save the JSON key path in .env:  GOOGLE_SERVICE_ACCOUNT_JSON=/path/to/key.json
  4. Create a Google Sheet, share it with the service account email (Editor)
  5. Copy the sheet ID from the URL and add to .env: GOOGLE_SHEET_ID=<id>

Install: pip install gspread google-auth

All scripts call: sync_jobs(all_jobs)  at the end of a run.
"""

import os
import logging
from datetime import datetime

from dotenv import load_dotenv

load_dotenv()

# ── Column order (matches Looker Studio dashboard schema) ─────────────────────
COLUMNS = [
    "date_applied",       # YYYY-MM-DD
    "platform",           # LinkedIn | Handshake | Indeed
    "job_id",
    "title",
    "company",
    "location",
    "remote_type",        # remote | onsite | hybrid
    "url",
    "employment_type",    # full-time | internship | part-time
    "salary_min",
    "salary_max",
    "salary_unit",        # yr | hr | mo
    "skills_mentioned",
    "apply_type",         # easy_apply | external | unknown
    "status",             # applied | skipped_* | error | scraped_only
    "days_since_posted",
    "notes",
]

_sheet = None   # cached worksheet

log = logging.getLogger(__name__)


def _get_sheet():
    """Return (and cache) the 'Applications' worksheet."""
    global _sheet
    if _sheet is not None:
        return _sheet

    try:
        import gspread
        from google.oauth2.service_account import Credentials
    except ImportError:
        raise ImportError("Run: pip install gspread google-auth")

    key_path = os.getenv("GOOGLE_SERVICE_ACCOUNT_JSON")
    sheet_id = os.getenv("GOOGLE_SHEET_ID")
    if not key_path or not sheet_id:
        raise EnvironmentError(
            "Set GOOGLE_SERVICE_ACCOUNT_JSON and GOOGLE_SHEET_ID in .env"
        )

    creds = Credentials.from_service_account_file(
        key_path,
        scopes=["https://www.googleapis.com/auth/spreadsheets"],
    )
    gc = gspread.authorize(creds)
    spreadsheet = gc.open_by_key(sheet_id)

    # Get or create the Applications tab
    try:
        ws = spreadsheet.worksheet("Applications")
    except gspread.WorksheetNotFound:
        ws = spreadsheet.add_worksheet("Applications", rows=10000, cols=len(COLUMNS))
        ws.append_row(COLUMNS)   # write header
        log.info("Created 'Applications' worksheet with headers.")

    _sheet = ws
    return ws


def _existing_ids(ws) -> set[str]:
    """Return set of 'platform|job_id' already in the sheet (dedup guard)."""
    try:
        records = ws.get_all_records()
        return {f"{r.get('platform','')}|{r.get('job_id','')}" for r in records if r.get("job_id")}
    except Exception:
        return set()


def _job_to_row(job: dict, platform: str) -> list:
    """Map a job dict (any script format) to the unified COLUMNS row."""
    # Normalise location field (scripts use different key names)
    location = (
        job.get("location_raw") or
        job.get("location") or
        job.get("city") or ""
    )
    return [
        job.get("timestamp", datetime.now().isoformat())[:10],   # date only
        platform,
        str(job.get("job_id", "")),
        job.get("title", ""),
        job.get("company", ""),
        location,
        job.get("remote_type", ""),
        job.get("url", ""),
        job.get("employment_type", ""),
        job.get("salary_min", ""),
        job.get("salary_max", ""),
        job.get("salary_unit", ""),
        job.get("skills_mentioned", ""),
        job.get("apply_type", ""),
        job.get("status", ""),
        job.get("days_since_posted", ""),
        job.get("notes", ""),
    ]


def sync_jobs(jobs: list[dict], platform: str, applied_only: bool = False):
    """
    Append new jobs to the Google Sheet.

    Args:
        jobs:         List of job dicts from any application script.
        platform:     "LinkedIn" | "Handshake" | "Indeed"
        applied_only: If True, only sync jobs with status="applied".
    """
    if not jobs:
        return
    try:
        ws = _get_sheet()
        existing = _existing_ids(ws)

        rows = []
        for job in jobs:
            if applied_only and job.get("status") != "applied":
                continue
            uid = f"{platform}|{job.get('job_id', '')}"
            if uid in existing:
                continue   # already synced
            rows.append(_job_to_row(job, platform))

        if rows:
            ws.append_rows(rows, value_input_option="USER_ENTERED")
            log.info(f"[gsheets] Synced {len(rows)} new rows to Google Sheets ({platform})")
        else:
            log.info("[gsheets] No new rows to sync.")
    except Exception as e:
        log.warning(f"[gsheets] Sync failed (non-fatal): {e}")

File: v5/lib/gsheets.mjs
// v5/lib/gsheets.mjs — Node port of library/gsheets.py.
//
// Appends new job rows to the "Applications" tab of a Google Sheet, matching
// the COLUMNS schema used by every other script in the suite. Dedupe key is
// "platform|job_id" — same as the Python version. Uses the `googleapis`
// package with a service-account JSON. If env vars or the package are
// missing, `syncJobs()` returns a soft-skip result rather than throwing,
// so the scraper still completes successfully without sheets configured.
//
// Required env (loaded via lib/config.mjs):
//   GOOGLE_SERVICE_ACCOUNT_JSON  — absolute path to service-account JSON key
//   GOOGLE_SHEET_ID              — the sheet ID (from its URL)

import fs from 'node:fs';
import { env } from './config.mjs';

const COLUMNS = [
  'date_applied', 'platform', 'job_id', 'title', 'company', 'location',
  'remote_type', 'url', 'employment_type', 'salary_min', 'salary_max',
  'salary_unit', 'skills_mentioned', 'apply_type', 'status',
  'days_since_posted', 'notes',
];

const SHEET_NAME = 'Applications';

function jobToRow(job, platform) {
  const location = job.location_raw || job.location || job.city || '';
  const ts = job.timestamp || new Date().toISOString();
  return [
    ts.slice(0, 10),                            // date_applied (YYYY-MM-DD)
    platform,
    String(job.job_id || ''),
    job.title || '',
    job.company || '',
    location,
    job.remote_type || '',
    job.url || '',
    job.employment_type || '',
    job.salary_min || '',
    job.salary_max || '',
    job.salary_unit || '',
    job.skills_mentioned || '',
    job.apply_type || '',
    job.status || '',
    job.days_since_posted ?? '',
    job.note || job.notes || '',
  ];
}

/** Lazy-load googleapis. Returns null if not installed. */
async function loadSheetsClient(serviceAccountJsonPath) {
  let google;
  try {
    ({ google } = await import('googleapis'));
  } catch {
    return null;
  }
  const auth = new google.auth.GoogleAuth({
    keyFile: serviceAccountJsonPath,
    scopes: ['https://www.googleapis.com/auth/spreadsheets'],
  });
  return google.sheets({ version: 'v4', auth });
}

/** Read existing rows and return Set of "platform|job_id" already synced. */
async function existingIds(sheets, sheetId) {
  try {
    const res = await sheets.spreadsheets.values.get({
      spreadsheetId: sheetId,
      range: `${SHEET_NAME}!A:Q`,
    });
    const rows = res.data.values || [];
    if (rows.length === 0) return new Set();
    const header = rows[0];
    const platformIdx = header.indexOf('platform');
    const jobIdIdx = header.indexOf('job_id');
    if (platformIdx < 0 || jobIdIdx < 0) return new Set();
    return new Set(
      rows.slice(1)
        .filter((r) => r[jobIdIdx])
        .map((r) => `${r[platformIdx] || ''}|${r[jobIdIdx]}`),
    );
  } catch {
    return new Set();
  }
}

/** Ensure the Applications tab exists with a header row. */
async function ensureSheet(sheets, sheetId, log) {
  try {
    const meta = await sheets.spreadsheets.get({ spreadsheetId: sheetId });
    const exists = (meta.data.sheets || []).some((s) => s.properties?.title === SHEET_NAME);
    if (exists) return;
    await sheets.spreadsheets.batchUpdate({
      spreadsheetId: sheetId,
      requestBody: { requests: [{ addSheet: { properties: { title: SHEET_NAME } } }] },
    });
    await sheets.spreadsheets.values.update({
      spreadsheetId: sheetId,
      range: `${SHEET_NAME}!A1`,
      valueInputOption: 'USER_ENTERED',
      requestBody: { values: [COLUMNS] },
    });
    log?.info?.(`gsheets: created '${SHEET_NAME}' worksheet with headers`);
  } catch (e) {
    throw new Error(`ensureSheet failed: ${e.message}`);
  }
}

/**
 * Append new jobs to the configured Google Sheet.
 *
 * @param {Array<object>} jobs       records from RunOutput.records()
 * @param {string} platform           "LinkedIn" | "Handshake" | "Indeed" | ...
 * @param {object} [opts]
 * @param {boolean} [opts.appliedOnly]  default false — only sync status=applied
 * @param {object}  [opts.log]          logger with .info/.warn
 * @returns {Promise<{appended: number, skipped: string}>}
 */
export async function syncJobs(jobs, platform, opts = {}) {
  const log = opts.log;
  if (!jobs || jobs.length === 0) return { appended: 0, skipped: 'no_jobs' };
  if (!env.GOOGLE_SERVICE_ACCOUNT_JSON || !env.GOOGLE_SHEET_ID) {
    log?.info?.('gsheets: GOOGLE_SERVICE_ACCOUNT_JSON / GOOGLE_SHEET_ID not set — skipping sync');
    return { appended: 0, skipped: 'no_credentials' };
  }
  if (!fs.existsSync(env.GOOGLE_SERVICE_ACCOUNT_JSON)) {
    log?.warn?.(`gsheets: service-account file not found at ${env.GOOGLE_SERVICE_ACCOUNT_JSON}`);
    return { appended: 0, skipped: 'no_credentials_file' };
  }

  const sheets = await loadSheetsClient(env.GOOGLE_SERVICE_ACCOUNT_JSON);
  if (!sheets) {
    log?.warn?.('gsheets: googleapis package not installed — run `npm install googleapis` to enable sync');
    return { appended: 0, skipped: 'package_missing' };
  }

  await ensureSheet(sheets, env.GOOGLE_SHEET_ID, log);
  const existing = await existingIds(sheets, env.GOOGLE_SHEET_ID);

  const rows = [];
  for (const job of jobs) {
    if (opts.appliedOnly && job.status !== 'applied') continue;
    const uid = `${platform}|${job.job_id || ''}`;
    if (existing.has(uid)) continue;
    rows.push(jobToRow(job, platform));
  }
  if (rows.length === 0) return { appended: 0, skipped: 'all_dupes' };

  await sheets.spreadsheets.values.append({
    spreadsheetId: env.GOOGLE_SHEET_ID,
    range: `${SHEET_NAME}!A:Q`,
    valueInputOption: 'USER_ENTERED',
    requestBody: { values: rows },
  });
  return { appended: rows.length, skipped: '' };
}

export { COLUMNS, jobToRow };
